File: Jan-2023/Java/ConcatenatedWords.java
// 472. Concatenated Words
class Solution {
	Set<String> set = new HashSet<>();
	public List<String> findAllConcatenatedWordsInADict(String[] words) {
		List<String> result = new LinkedList<>();
		//add them to set and check them later
		for (String s : words) {
			set.add(s);
		}
		for (int i = 0; i < words.length; i++) {
			String w = words[i];
			if (isPresent(w)) {
				result.add(w);
			}
		}
		return result;
	}
	public boolean isPresent(String word) {
		for (int i = 0; i < word.length(); i++) {
			//cut the string into two seperate parts and check if they contain within the hashset
			//hashset peforms operation in O(1) in avearge case so it is better than linkedlist which searches in O(n) 
			String prefix = word.subString(0, i + 1);
			String suffix = word.subString(i + 1, word.length());
			if ((set.contains(prefix) && set.contains(suffix)) || set.contains(prefix) && isPresent(suffix)) {
				return true;
			}
		}
		return false;
	}
}

File: Jan-2023/C++/ConcatenatedWords.cxx


// 472. Concatenated Words
class Solution {
    
public:
    unordered_set<string> set;
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        for(auto s: words){
      	//add them to set and check them later
            set.insert(s);

        }
        vector<string> result;
        for(int i=0; i<words.size(); i++){
            string word = words[i];
            if(isPresent(word)){
                result.push_back(word);
              }
            }
        return result;
    }
    
    bool isPresent(string word){
        for(int i=0;i<word.length();i++){
            	//cut the string into two seperate parts and check if they contain within the hashset
			//hashset peforms operation in O(1) in avearge case so it is better than linkedlist which searches in O(n) 
            string prefix = word.substr(0,i+1); 
            string suffix = word.substr(i+1);
            if((set.count(prefix) && set.count(suffix)) || (set.count(prefix) && isPresent(suffix))){
                return true;
                }
            
          }
        return false;
     }
};
